src/visitors/derivative_original_visitor.hpp

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2023 Blue Brain Project, EPFL.
+ * See the top-level LICENSE file for details.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#pragma once
+
+/**
+ * \file
+ * \brief \copybrief nmodl::visitor::DerivativeOriginalVisitor
+ */
+
+#include "symtab/decl.hpp"
+#include "visitors/ast_visitor.hpp"
+#include <string>
+
+namespace nmodl {
+namespace visitor {
+
+/**
+ * \addtogroup visitor_classes
+ * \{
+ */
+
+/**
+ * \class DerivativeOriginalVisitor
+ * \brief Make a copy of the `DERIVATIVE` block (if it exists), and insert back as
+ * `DERIVATIVE_ORIGINAL_FUNCTION` block.
+ *
+ * If \ref SympySolverVisitor runs successfully, it replaces the original
+ * solution. This block is inserted before that to prevent losing access to
+ * information about the block.
+ */
+class CvodeVisitor: public AstVisitor {
+  private:
+    /// The copy of the derivative block we are solving
+    std::shared_ptr<ast::DerivativeBlock> der_block = nullptr;
+
+    /// true while visiting differential equation
+    bool differential_equation = false;
```

Comment:
Are the names for `differential_equation` and `derivative_block` idiomatic? Could they have a better name? Something the hints better at what they are: `inside_*`, `beneath_*`, `is_*`, `printing_*`